clients/src/main/java/org/apache/kafka/common/requests/ProduceRequest.java

```diff
@@ -133,6 +144,25 @@ public Struct toStruct() {
     }
 
     @Override
+    public String toString() {
+        // Store it in a local variable to protect against concurrent updates
+        Map<TopicPartition, MemoryRecords> partitionRecords = this.partitionRecords;
+
+        StringBuilder bld = new StringBuilder();
+        bld.append("(type=ProduceRequest")
+                .append(", acks=").append(acks)
+                .append(", timeout=").append(timeout);
+
+        if (partitionRecords == null)
+            bld.append(", partitions=").append(Utils.mkString(partitions, ","));
+        else
+            bld.append(", partitionRecords=(").append(Utils.mkString(partitionRecords));
```

Comment:
Does this print out the record contents as well? If so, that seems very excessive.

I think the old ProduceRequest had it right: print out the various fields and on initialization of ProduceRequest, store a size-per-partition mapping so that even after clearing records, we can log the size-per-partition.